src/sources/postgresql_metrics.rs

```diff
@@ -0,0 +1,851 @@
+        );
+        let namespace = Some(self.namespace.clone()).filter(|namespace| !namespace.is_empty());
+
+        let mut sources = try_join_all(self.endpoints.iter().map(|endpoint| {
+            PostgresqlMetrics::new(
+                endpoint.clone(),
+                datname_filter.clone(),
+                namespace.clone(),
+                self.tls.clone(),
+            )
+        }))
+        .await?;
+
+        let mut out =
+            out.sink_map_err(|error| error!(message = "Error sending postgresql metrics.", %error));
+
+        let duration = time::Duration::from_secs(self.scrape_interval_secs);
+        Ok(Box::pin(async move {
+            let mut interval = time::interval(duration).take_until(shutdown);
+            while interval.next().await.is_some() {
+                let start = Instant::now();
+                let metrics = join_all(sources.iter_mut().map(|source| source.collect())).await;
+                emit!(PostgresqlMetricsCollectCompleted {
+                    start,
+                    end: Instant::now()
+                });
+
+                let mut stream = stream::iter(metrics).flatten().map(Event::Metric).map(Ok);
+                out.send_all(&mut stream).await?;
+            }
+
+            Ok(())
+        }))
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Metric
+    }
+
+    fn source_type(&self) -> &'static str {
+        "postgresql_metrics"
+    }
+}
+
+#[derive(Debug, Clone)]
+enum DatnameFilter {
+    Include {
+        need_null: bool,
+        databases: Vec<String>,
+    },
+    Exclude {
+        need_null: bool,
+        databases: Vec<String>,
+    },
+    None,
+}
+
+impl DatnameFilter {
+    fn new(include: Vec<String>, exclude: Vec<String>) -> Self {
+        let exclude = exclude.into_iter().collect::<HashSet<_>>();
+        let include = include
+            .into_iter()
+            .filter(|name| !exclude.contains(name))
+            .collect::<HashSet<_>>();
+
+        if !include.is_empty() {
+            let (need_null, databases) = Self::remove_empty(include);
+            return Self::Include {
+                need_null,
+                databases,
+            };
+        }
+
+        if !exclude.is_empty() {
+            let (need_null, databases) = Self::remove_empty(exclude);
+            return Self::Exclude {
+                need_null,
+                databases,
+            };
+        }
+
+        Self::None
+    }
+
+    fn remove_empty(mut names: HashSet<String>) -> (bool, Vec<String>) {
+        let null = names.remove(&"".to_owned());
+        (null, names.into_iter().collect())
+    }
+
+    async fn pg_stat_database(&self, client: &Client) -> Result<Vec<Row>, PgError> {
+        let mut conditions = "SELECT * FROM pg_stat_database".to_owned();
+        match self {
+            Self::Include {
+                need_null,
+                databases,
+            } => {
+                conditions += " WHERE datname = ANY($1)";
```

Comment:
I still think it'd be useful to support wildcards for the include/exclude filters. This would let me write things like:

```
exclude = ["template*"]
```

To exclude all template dbs. I think it could be translated into a `LIKE` query here.

Reply:
Hm, I think it would be cool, don't you mind if I add this in separate PR? I'd prefer to land this without adding optional enhancements.

Reply:
:+1: I'm fine with that as a follow-up issue.